libjcat/jcat-bt-util.c

```diff
@@ -144,3 +146,405 @@ jcat_bt_generate_key_pair(const gchar *keyname,
+	}
+	if (*parsed_key_name) {
+		g_free(*parsed_key_name);
+	}
+	*parsed_key_name = g_strndup((const gchar *)data, next - data);
+	key_name_size = next - data;
+
+	data += key_name_size + 1;
+	size -= key_name_size + 1;
+
+	/* The next field is the key hash. */
+	next = memchr(data, '+', size);
+	if (next == NULL) {
+		g_set_error_literal(error,
+				    G_IO_ERROR,
+				    G_IO_ERROR_FAILED,
+				    "cannot find end of key hash");
+		return FALSE;
+	}
+	if (next != data + key_hash_size) {
+		g_set_error(error,
+			    G_IO_ERROR,
+			    G_IO_ERROR_FAILED,
+			    "incorrect size for key hash; want %zu got %ld",
+			    key_hash_size,
+			    next - data);
+		return FALSE;
+	}
+	key_hash = data;
+	for (; data != next; ++data, --size) {
+		if (!((*data >= '0' && *data <= '9') || (*data >= 'a' && *data <= 'f') ||
+		      (*data >= 'A' && *data <= 'F'))) {
+			g_set_error_literal(error,
+					    G_IO_ERROR,
+					    G_IO_ERROR_FAILED,
+					    "key hash must be hexadecimal");
+			return FALSE;
+		}
+	}
+
+	++data;
+	--size;
+
+	/* The final field is the base64-encoded key content. Need to be NULL-terminated.*/
+	base64_private = g_string_new_len((const gchar *)data, size);
+	for (; size > 0; ++data, --size) {
+		if (!g_ascii_isalnum(*data) && *data != '+' && *data != '/') {
+			g_set_error_literal(error,
+					    G_IO_ERROR,
+					    G_IO_ERROR_FAILED,
+					    "incorrect base64 encoding");
+			return FALSE;
+		}
+	}
+	private = g_base64_decode(base64_private->str, &private_len);
+	if (private_len != 1 + ED25519_KEY_SIZE) {
+		g_set_error_literal(error,
+				    G_IO_ERROR,
+				    G_IO_ERROR_FAILED,
+				    "unexpected size for private key");
+		return FALSE;
+	}
+	if (*private != 1) {
+		g_set_error_literal(error,
+				    G_IO_ERROR,
+				    G_IO_ERROR_FAILED,
+				    "unsupported key type; want Ed25519");
+		return FALSE;
+	}
+
+	_g_set_byte_array(parsed_private_key,
+			  g_byte_array_append(g_byte_array_new(), private + 1, ED25519_KEY_SIZE));
+	_g_set_byte_array(parsed_public_key,
+			  g_byte_array_set_size(g_byte_array_new(), ED25519_KEY_SIZE));
+
+	ed25519_sha512_public_key((*parsed_public_key)->data, (*parsed_private_key)->data);
+
+	/* Check the key hash */
+	public_key_for_validation = g_byte_array_new();
+	public_key_for_validation = g_byte_array_append(public_key_for_validation, &ed25519_id, 1);
+	public_key_for_validation = g_byte_array_append(public_key_for_validation,
+							(*parsed_public_key)->data,
+							(*parsed_public_key)->len);
+	expected_key_hash = generate_key_hash(*parsed_key_name, public_key_for_validation);
+	if (memcmp(expected_key_hash, key_hash, key_hash_size) != 0) {
+		g_set_error(
+		    error,
+		    G_IO_ERROR,
+		    G_IO_ERROR_FAILED,
+		    "Mismatching key hash: provided key hash does not match expected key hash %s",
+		    expected_key_hash);
+		return FALSE;
+	}
+	if (*parsed_key_hash) {
+		g_free(*parsed_key_hash);
+	}
+	*parsed_key_hash = g_new(char, 1 + key_hash_size / 2);
```

Comment:
g_new0 pls -- then it's not filled with junk

Reply:
Done.